gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/GraphManager.java

```diff
@@ -21,139 +21,98 @@
     /**
      * Get the {@link Graph} and {@link TraversalSource} list as a set of bindings.
      */
-    public Bindings getAsBindings() {
-        final Bindings bindings = new SimpleBindings();
-        graphs.forEach(bindings::put);
-        traversalSources.forEach(bindings::put);
-        return bindings;
-    }
+
+    /**
+     * Add {@link TraversalSource} ts with name {@link String} tsName to
+     * {@link Map<String, TraversalSource>} returned by call to getTraversalSources()
+     */
+    public void addTraversalSource(final String tsName, final TraversalSource ts);
+
+    public Bindings getAsBindings();
 
     /**
      * Rollback transactions across all {@link Graph} objects.
      */
-    public void rollbackAll() {
-        graphs.entrySet().forEach(e -> {
-            final Graph graph = e.getValue();
-            if (graph.features().graph().supportsTransactions() && graph.tx().isOpen())
-                graph.tx().rollback();
-        });
-    }
+    public void rollbackAll();
 
     /**
      * Selectively rollback transactions on the specified graphs or the graphs of traversal sources.
      */
-    public void rollback(final Set<String> graphSourceNamesToCloseTxOn) {
-        closeTx(graphSourceNamesToCloseTxOn, Transaction.Status.ROLLBACK);
-    }
+    public void rollback(final Set<String> graphSourceNamesToCloseTxOn);
 
     /**
      * Commit transactions across all {@link Graph} objects.
      */
-    public void commitAll() {
-        graphs.entrySet().forEach(e -> {
-            final Graph graph = e.getValue();
-            if (graph.features().graph().supportsTransactions() && graph.tx().isOpen())
-                graph.tx().commit();
-        });
-    }
+    public void commitAll();
 
     /**
      * Selectively commit transactions on the specified graphs or the graphs of traversal sources.
      */
-    public void commit(final Set<String> graphSourceNamesToCloseTxOn) {
-        closeTx(graphSourceNamesToCloseTxOn, Transaction.Status.COMMIT);
-    }
+    public void commit(final Set<String> graphSourceNamesToCloseTxOn);
+
+    /**
+     * Implementation that allows for custom graph-opening implementations; if the {@link Map}
+     * tracking graph references has a {@link Graph} object corresponding to the {@link String} graphName,
+     * then we return that {@link Graph}-- otherwise, we use the custom {@link Supplier} to instantiate a
+     * a new {@link Graph}, add it to the {@link Map} tracking graph references, and return said {@link Graph}.
+     */
+    public Graph openGraph(final String graphName, final Supplier<Graph> supplier);
+
+    /**
+     * Implementation that allows for custom graph-closing implementations; it is up to the implementor
+     * to decide if this method should also remove the {@link Graph} graph from the {@link Map}
+     * tracking {@link Graph} references (and how it would do so).
+     */
+    public void closeGraph(final Graph graph) throws Exception;
```

Comment:
I've made a number of comments about `closeGraph()` so another one shouldn't be surprising :smile: - I can sorta understand the requirement here as the symmetrical method to `openGraph()` and the idea that the component interacting with the manager won't keep track of the graph names and may only have the `Graph` instance. So, I'm that far along....

So these would be the current thoughts/concerns:

1. The javadoc makes this implementation have a pretty wide open scope leaving it up to the implementation to choose whether or not the `Graph` is removed from the "registry" (i.e. control of the `GraphManager`) - you mention `Map` specifically but the registry might not use that data structure specifically I suppose.  Under what circumstance would an interacting component call `closeGraph()` but not want the `Graph` removed from the registry? To leave it there would allow callers to Gremlin Server to get graphs that were in a "closed" state which doesn't seem right. 
2. Piggy-backing on that last point, if you go down to `DefaultGraphManager`, I think that it should look to remove the instance from the registry. Again, I'm lost for a scenario where you would want a closed `Graph` instance available to Gremlin Server requests.
3. I still think this method is really just an overload of `removeGraph(String)` and should be named as such: `Graph removeGraph(Graph)`. A bit weird as you already have the `Graph` instance and are just returning it back, but it would let the API be more fluent. I think @pluradj implied the same thing in a different comment.